### ISSUE TITLE
KJ_IREQUIRE fails to compile on MSVC with /Zc:preprocessor enabled

### ISSUE THREAD

**Issue by nyanpasu64** [capnproto/capnproto#1307]
The `KJ_IREQUIRE` macro currently branches based on MSVC vs. other compilers in its attempt to make the leading comma in a `__VA_ARGS__` block conditional. It uses `"" #__VA_ARGS__, __VA_ARGS__` on MSVC, and `#__VA_ARGS__, ##__VA_ARGS__` elsewhere. (I'm not sure what either does, are you stringifying `__VA_ARGS__`? In any case, `__VA_ARGS__` commas are a hard problem solved differently across compilers and versions, and I understand the difficulties and edge cases.)

https://github.com/capnproto/capnproto/blob/master/c++/src/kj/common.h#L297-L313

However, on recent versions of MSVC 2019, the [`/Zc:preprocessor` flag](https://devblogs.microsoft.com/cppblog/announcing-full-support-for-a-c-c-conformant-preprocessor-in-msvc/) makes its preprocessor act more similarly to GCC, not omitting the leading comma and instead depending on `##__VA_ARGS__` or `__VA_OPT__` to drop the leading comma. I'm not sure exactly what the difference is though. In any case, with `/Zc:preprocessor` passed, the MSVC case fails to compile and only the non-MSVC case works.

The blog says:

> The conformant preprocessor can be tested by checking if the macro _MSVC_TRADITIONAL is defined and set to 0.

And subsequently presents the incorrect code for testing if the conformant preprocessor is turned off:

```cpp
#if defined(_MSVC_TRADITIONAL) && _MSVC_TRADITIONAL
// old behavior
#else
// new behavior
#endif
```

This triggers `new behavior` both on old MSVC versions before `_MSVC_TRADITIONAL` existed, *and* on newer versions with `_MSVC_TRADITIONAL=0`. And this is *after* a commenter "corrected" the original blog post. Boolean logic is hard, De Morgan's laws are hard... I think Microsoft should've added a `__MSVC_NEW_PREPROCESSOR` flag instead, to simplify testing for the new preprocessor (1) and old preprocessor (unset or 0).

In any case, to to test for new behavior, you need `#if defined(_MSVC_TRADITIONAL) && !_MSVC_TRADITIONAL`, and to test for old behavior, you need `#if !defined(_MSVC_TRADITIONAL) || _MSVC_TRADITIONAL`. I think I got it right? Double negatives are hard.

In summary, the most concise fix is to switch to `#if _MSC_VER && !defined(__clang__) && (!defined(_MSVC_TRADITIONAL) || _MSVC_TRADITIONAL)`, but this turns out to be a rather long line of text.

There may be other ways to write the command, maybe just `_MSVC_TRADITIONAL` but this falsely detects MSVC before the macro [was introduced](https://docs.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=msvc-160#:~:text=_msvc_traditional%3A) (pre-VS2017 and older versions of VS2017 and VS2019) as "new preprocessor".

One alternative approach that doesn't introduce a new layer of nesting is:

```cpp
#ifndef _MSVC_TRADITIONAL
#define _MSVC_TRADITIONAL 1
#endif
#if _MSC_VER && !defined(__clang__) && _MSVC_TRADITIONAL
```

**Comment by kentonv** [comment]
Interesting, so it sounds like we potentially need to use the GCC syntax based on macros. I suppose the macros in `kj/debug.h` probably need to pay attention to this as well, and probably anywhere in the codebase that uses `__VA_ARGS__`.

(And yes, `#__VA_ARGS__` is meant to stringify the args, while `##__VA_ARGS__` is the GCC syntax to expand the args or remove the preceding comma if they are empty.)
